src/state_machine.rs

```diff
@@ -322,6 +336,75 @@ impl StateMachine {
         }
     }
 
+    /// Query for a result, or yield: Err(NothingAvailable), Err(Disconnected).
+    #[cfg(feature = "use-mock-crust")]
+    pub fn try_step(&mut self, outbox: &mut EventBox) -> Result<(), TryRecvError> {
+        use itertools::Itertools;
+        use maidsafe_utilities::SeededRng;
+        use rand::Rng;
+        use std::borrow::Borrow;
+
+        if self.is_running {
+            let mut events = Vec::new();
+            while let Ok(category) = self.category_rx.try_recv() {
+                match category {
+                    MaidSafeEventCategory::Routing => {
+                        if let Ok(action) = self.action_rx.try_recv() {
+                            events.push(EventType::Action(Box::new(action)));
+                        } else {
+                            return Ok(self.apply_transition(Transition::Terminate, outbox));
+                        }
+                    }
+                    MaidSafeEventCategory::Crust => {
+                        if let Ok(crust_event) = self.crust_rx.try_recv() {
+                            events.push(EventType::CrustEvent(crust_event));
+                        } else {
+                            return Ok(self.apply_transition(Transition::Terminate, outbox));
+                        }
+                    }
+                }
+            }
+
+            let timed_out_events = self.state
+                .get_timed_out_tokens()
+                .iter()
+                .map(|token| EventType::Action(Box::new(Action::Timeout(*token))))
+                .collect_vec();
+
+            if !events.is_empty() {
+                for event in timed_out_events {
+                    let pos = SeededRng::thread_rng().gen_range(0, events.len());
+                    events.insert(pos, event);
+                }
+            } else {
+                events.extend(timed_out_events);
+            }
+            self.events.extend(events);
+
+            let is_all_timed_out = self.events
+                .iter()
+                .all(|event| match *event {
+                         EventType::Action(ref action) => {
+                             match *action.borrow() {
+                                 Action::Timeout(_) => true,
+                                 _ => false,
+                             }
+                         }
+                         _ => false,
+                     });
```

Comment:
self.events may contain events from previous. so even `events` is empty for this call, there is still chance self.events contains non-time-out events.